src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1320,14 +1318,30 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
 
 	DeleteNodeRow(workerNode->workerName, nodePort);
 
+	RemoveOldShardPlacementForNodeGroup(workerNode->groupId);
```

Comment:
I thought about this a bit and couldn't find a case where this could be problematic, but it can still be problematic. 

One difference is that deferred drop is not transactional, however this one is transactional. And here we are only updating pg_dist_placement, we are not actually dropping any data

Reply:
I can test this but I think one will error out. It seems like this is the same as `what happens when we run DeleteShardPlacementRow concurrently` because that is the only common block for the two

Reply:
Sure I don't see any downside adding this lock to prevent that but since we are not dropping any actual data here, why do you think that is problematic(I mean getting the error)?

Reply:
You can trigger deferred drop manually with:
```
CREATE OR REPLACE FUNCTION master_defer_delete_shards()
    RETURNS int
    LANGUAGE C STRICT
    AS 'citus', $$master_defer_delete_shards$$;
COMMENT ON FUNCTION master_defer_delete_shards()
    IS 'remove orphaned shards';
SELECT master_defer_delete_shards();
```

and see that it gets blocked if  `begin; select citus_remove_node(...)` is active.

It seems that deferred drop does go ahead with dropping the shard if the node is still up. If the citus_remove_node is removing the record, it blocks on the row-level locks, but can later proceed regardless of whether the citus_remove_node commits.

However, that does mean that a citus_remove_node could block the maintenance daemon so it's better to use LockPlacementCleanup, such that the maintenance daemon backs off if the lock is taken.